bitcoin/src/psbt/mod.rs

```diff
@@ -201,135 +256,229 @@ impl Psbt {
     /// ## Panics
     ///
     /// The function panics if the length of transaction inputs is not equal to the length of PSBT inputs.
-    pub fn iter_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
-        assert_eq!(self.inputs.len(), self.unsigned_tx.input.len());
-        self.unsigned_tx.input.iter().zip(&self.inputs).map(|(tx_input, psbt_input)| {
-            match (&psbt_input.witness_utxo, &psbt_input.non_witness_utxo) {
-                (Some(witness_utxo), _) => Ok(witness_utxo),
-                (None, Some(non_witness_utxo)) => {
-                    let vout = tx_input.previous_output.vout as usize;
-                    non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
-                }
-                (None, None) => Err(Error::MissingUtxo),
+    pub fn iter_funding_utxos(&self) -> Box<dyn Iterator<Item = Result<&TxOut, Error>>> {
+        let inner = &self.inner;
+        match inner.version {
+            Version::PsbtV0 => {
+                let unsigned_tx = inner.unsigned_tx.as_ref().unwrap();
+                assert_eq!(inner.inputs.len(), unsigned_tx.input.len());
+                Box::new(unsigned_tx.input.iter().zip(&inner.inputs).map(|(tx_input, psbt_input)| {
```

Comment:
Why did you need to do Box dyn here? Is the allocation unavoidable here?

Reply:
As I checked, it looks like each match case needs to return the same thing. In the first case (PsbtV0), it returns a `Map<Zip<Iter<'_, TxIn>, Iter<'_, Input>>, impl Fn((&TxIn, &Input)) -> Result<&TxOut, Error>>`, whereas the second case returns a `Map<Iter<'_, Input>, impl Fn(&Input) -> Result<&TxOut, Error>>`. This was throwing an error and suggesting adding Box dyn here. I tried different possible ways but could not bypass the error. Will try exploring better ways.

Reply:
You may be able to use a `&dyn` rather than a `Box<dyn>`, which avoids the allocation.

The other way to handle this is to extract code into a generic utility function, but this is often pretty ugly and it often requires you move more code than you expect.